### ISSUE TITLE
dyld: Symbol not found: ___CFObjCIsCollectable

### ISSUE THREAD

**Issue by kirelagin** [NixOS/nixpkgs#45042]
This happens on macOS Mojave 10.14:

```haskell
{-# LANGUAGE OverloadedStrings #-}

import Control.Monad
import Data.Default.Class
import Network.HTTP.Req

main :: IO ()
main = runReq def $
  void $ req GET (https "api.github.com" /: "meta") NoReqBody ignoreResponse mempty
```

```
› nix-shell -p 'haskellPackages.ghcWithPackages (hp: [ hp.req ])' --run 'runhaskell Main.hs'
dyld: Symbol not found: ___CFObjCIsCollectable
  Referenced from: /nix/store/wkqjqwyfi1nn2hg9k0as03lpcmalqxzj-libsecurity_utilities-osx-10.7.5/lib/libsecurity_utilities.0.dylib
  Expected in: /System/Library/Frameworks/CoreFoundation.framework/Versions/A/CoreFoundation
 in /nix/store/wkqjqwyfi1nn2hg9k0as03lpcmalqxzj-libsecurity_utilities-osx-10.7.5/lib/libsecurity_utilities.0.dylib
Main.hs: VanillaHttpException (HttpExceptionRequest Request {
  host                 = "api.github.com"
  port                 = 443
  secure               = True
  requestHeaders       = []
  path                 = "/meta"
  queryString          = ""
  method               = "GET"
  proxy                = Nothing
  rawBody              = False
  redirectCount        = 10
  responseTimeout      = ResponseTimeoutDefault
  requestVersion       = HTTP/1.1
}
 (InternalException (HandshakeFailed (Error_Protocol ("certificate has unknown CA",True,UnknownCa)))))
```

(the expected result is a 403.)

I am not sure if it is a Mojave-specific issue as I currently don’t have a previous version of macOS to test with, but I’ll ping #42719 just in case.

**Comment by kirelagin** [comment]
```
› objdump /System/Library/Frameworks/CoreFoundation.framework/Versions/A/CoreFoundation -macho -exports-trie | grep ___CFObjCIsCollectable                                                                                                                                                                  <1>
0x00594E50  ___CFObjCIsCollectable
```
🤔

**Comment by LnL7** [comment]
Hmm, I was expecting this to be the problem but that doesn't seem to be it. I was expecting our version not to have it while libsecurity is expecting it to be there.

```
$ nm -g /nix/store/36ph03q316x86rdyvx63ha2kl51f84dm-CF-osx-10.10.5/Library/Frameworks/CoreFoundation.framework/Versions/A/CoreFoundation | grep ___CFObjCIsCollectable
00000000000f6970 S ___CFObjCIsCollectable
$ nm -g /System/Library/Frameworks/CoreFoundation.framework/Versions/A/CoreFoundation | grep ___CFObjCIsCollectable
00000000005e68d0 S ___CFObjCIsCollectable
```

**Comment by kirelagin** [comment]
Got access to macOS 10.13.6 and it works there, so, yes, a Mojave regression.

**Comment by jhenahan** [comment]
Speculating: There appear to be some new SSL-related headers that aren't being copied into our libsecurity:

    /System/Library/Frameworks/Security.framework/Versions/A/Headers/SecProtocolMetadata.h
    /System/Library/Frameworks/Security.framework/Versions/A/Headers/SecProtocolOptions.h

I'm not sure how to add those to my local `nixpkgs` checkout to test if that'll fix it, but it's something.

**Comment by jhenahan** [comment]
Due to a suggestion at https://github.com/NixOS/nixpkgs/blob/7736337916538aab67a00739c53b8434da773a3b/doc/functions.xml#L636 about this specific error message, I'm attempting a GHC build with `cacert` added to `nativeBuildInputs` to see if that gets things any further along. Will update if/when it finishes.

**Comment by cransom** [comment]
I tried to follow the story here and I didn't see a resolution, so I assume this is still broken. I'm on the current bleeding master and trying to build `haskellPackages.dhall_1_17_0` and I'm stumbling into this as well:
```
/private/tmp/nix-build-dhall-1.17.0.drv-0/dhall-1.17.0/src/Dhall/Tutorial.hs:440: failure in expression `input auto "https://raw.githubusercontent.com/dhall-lang/dhall-haskell/18e4e9a18dc53271146df3ccf5b4177c3552236b/examples/True" :: IO Bool'
expected: True
 but got: dyld: Symbol not found: ___CFObjCIsCollectable
            Referenced from: /nix/store/79x43vkf7r0hhyd7h6xy7kipyv9j5kj8-libsecurity_utilities-osx-10.7.5/lib/libsecurity_utilities.0.dylib
            Expected in: /System/Library/Frameworks/CoreFoundation.framework/Versions/A/CoreFoundation
           in /nix/store/79x43vkf7r0hhyd7h6xy7kipyv9j5kj8-libsecurity_utilities-osx-10.7.5/lib/libsecurity_utilities.0.dylib
          dyld: Symbol not found: ___CFObjCIsCollectable
            Referenced from: /nix/store/79x43vkf7r0hhyd7h6xy7kipyv9j5kj8-libsecurity_utilities-osx-10.7.5/lib/libsecurity_utilities.0.dylib
            Expected in: /System/Library/Frameworks/CoreFoundation.framework/Versions/A/CoreFoundation
           in /nix/store/79x43vkf7r0hhyd7h6xy7kipyv9j5kj8-libsecurity_utilities-osx-10.7.5/lib/libsecurity_utilities.0.dylib
<snip>
```
Darwin is greek to me, but from the previous commands, it looks like the symbol isn't copied or at least is undefined?
```
0caseys-imac:~/git/nixpkgs% nm -g /nix/store/79x43vkf7r0hhyd7h6xy7kipyv9j5kj8-libsecurity_utilities-osx-10.7.5/lib/libsecurity_utilities.0.dylib | grep ___CFObjCIsCollectable
                 U ___CFObjCIsCollectable
0caseys-imac:~/git/nixpkgs% nm -g /System/Library/Frameworks/CoreFoundation.framework/Versions/A/CoreFoundation | grep ___CFObjCIsCollectable
0000000000595e60 S ___CFObjCIsCollectable
```
This would seem to line up with @LnL7's theory from https://github.com/NixOS/nixpkgs/issues/45042#issuecomment-413031009

**Comment by jhenahan** [comment]
@cransom For the immediate problem of `dhall`, you can override it with `dontCheck` to skip the tests. Network calls still won't work, but it'll at least be usable.

    dhall = dontCheck super.dhall_1_17_0;

**Comment by matthewbauer** [comment]
I think we should try using the impure Security framework for now. I looked at the issue and it's definitely not obvious what to do. Perhaps updating to the latest Apple open source release fixes it but just as likely could be something else.

/cc @copumpkin

**Comment by domenkozar** [comment]
Unless Apple has a separate channel for alphas/betas, we have the latest version from https://opensource.apple.com/tarballs/libsecurity_utilities/

**Comment by matthewbauer** [comment]
They’ve moved it all to security. Those are from like 10.9

**Comment by matthewbauer** [comment]
I opened https://github.com/NixOS/nixpkgs/pull/47676 to fix this (temporarily at least).

**Comment by eyeinsky** [comment]
@matthewbauer thank you for doing this!

**Comment by copumpkin** [comment]
@domenkozar no objections

**Comment by Ekleog** [comment]
(triage) If I read it correctly, this is currently solved by #47676?

**Comment by matthewbauer** [comment]
Yeah as well as we can until we can update Security.
